gcloud-java-contrib/gcloud-java-nio/src/test/java/com/google/cloud/storage/contrib/nio/FakeStorageRpc.java

```diff
@@ -0,0 +1,374 @@
+    //return new DefaultRpcBatch(storage);
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public StorageObject compose(Iterable<StorageObject> sources, StorageObject target,
+                               Map<Option, ?> targetOptions) throws StorageException {
+    return null;
+  }
+
+  @Override
+  public byte[] load(StorageObject storageObject, Map<Option, ?> options) throws StorageException {
+    String key = fullname(storageObject);
+    if (!contents.containsKey(key)) {
+      throw new StorageException(404, "File not found: " + key);
+    }
+    return contents.get(key);
+  }
+
+  @Override
+  public Tuple<String, byte[]> read(
+      StorageObject from, Map<Option, ?> options, long zposition, int zbytes)
+      throws StorageException {
+    potentiallyThrow(options);
+    String key = fullname(from);
+    if (!contents.containsKey(key)) {
+      throw new StorageException(404, "File not found: " + key);
+    }
+    long position = zposition;
+    int bytes = zbytes;
+    if (position < 0) {
+      position = 0;
+    }
+    byte[] full = contents.get(key);
+    if ((int) position + bytes > full.length) {
+      bytes = full.length - (int) position;
+    }
+    if (bytes <= 0) {
+      // special case: you're trying to read past the end
+      return Tuple.of("etag-goes-here", new byte[0]);
+    }
+    byte[] ret = new byte[bytes];
+    System.arraycopy(full, (int) position, ret, 0, bytes);
+    return Tuple.of("etag-goes-here", ret);
+  }
+
+  @Override
+  public String open(StorageObject object, Map<Option, ?> options) throws StorageException {
+    String key = fullname(object);
+    boolean mustNotExist = false;
+    for (Option option : options.keySet()) {
+      if (option instanceof StorageRpc.Option) {
+        // this is a bit of a hack, since we don't implement generations.
+        if ((StorageRpc.Option) option == Option.IF_GENERATION_MATCH
+            && ((Long) options.get(option)).longValue() == 0L) {
+          mustNotExist = true;
+        }
+      }
+    }
+    if (mustNotExist && metadata.containsKey(key)) {
+      throw new StorageException(new FileAlreadyExistsException(key));
+    }
+    metadata.put(key, object);
+
+    return fullname(object);
+  }
+
+  @Override
+  public void write(String uploadId, byte[] toWrite, int toWriteOffset, long destOffset,
+                    int length, boolean last) throws StorageException {
+    // this may have a lot more allocations than ideal, but it'll work.
+    byte[] bytes;
+    if (futureContents.containsKey(uploadId)) {
+      bytes = futureContents.get(uploadId);
+      if (bytes.length < length + destOffset) {
+        bytes = new byte[(int) (length + destOffset)];
+      }
+    } else {
+      bytes = new byte[(int) (length + destOffset)];
+    }
+    System.arraycopy(toWrite, toWriteOffset, bytes, (int) destOffset, length);
+    // we want to mimic the GCS behavior that file contents are only visible on close.
```

Comment:
This makes sense for javadoc and READMEs. Here I believe it's of little use as the comment is for the sake of contributors which I believe are well aware of what GCS means